File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 131127. 할인 행사/kotlin/Solution.kt
class Solution {
    fun solution(want: Array<String>, number: IntArray, discount: Array<String>): Int {
        val wantCounter = HashMap<String, Int>()
        val counter = HashMap<String, Int>()
        for (i in 0 until want.size) {
            wantCounter[want[i]] = number[i]
            counter[want[i]] = 0
        }
        for (item in discount) {
            if (!counter.containsKey(item)) counter[item] = 0
            if (!wantCounter.containsKey(item)) wantCounter[item] = 0
        }
        for (i in 0 until 10) counter[discount[i]] = counter[discount[i]]!! + 1

        var answer = 0
        if (wantCounter == counter) answer ++
        for (day in 10 until discount.size) {
            counter[discount[day-10]] = counter[discount[day-10]]!! - 1
            counter[discount[day]] = counter[discount[day]]!! + 1
            if (wantCounter == counter) answer ++
        }
        return answer
    }
}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 131127. 할인 행사/python/solution.py
def solution(want, number, discount):
    answer = 0
    want_counter = {} # 희망 구매 목록
    daily_counter = {} # 현재 수량

    # want-number 쌍을 순회하면서 목록 초기화
    for w, n in zip(want, number):
        want_counter[w] = n
        daily_counter[w] = 0

    # discount 를 순회하면서 목록 초기화
    for item in discount:
        if item not in want_counter:
            want_counter[item] = 0
        if item not in daily_counter:
            daily_counter[item] = 0

    # 날짜 경과하면서 그날 할인품목을 추가(첫날부터), 제거(10일차부터)
    # 그날 할인품목과 희망 할인품목 비교(9 일차부터)
    for day in range(len(discount)):
        daily_counter[discount[day]] += 1
        if day >= 10:
            daily_counter[discount[day - 10]] -= 1
        if day >= 9 and want_counter == daily_counter:
            answer += 1
    return answer
